File: listen.py
import os
import json
import pyaudio
import websockets
import asyncio
from dotenv import load_dotenv
from deepgram import DeepgramClient, DeepgramClientOptions, LiveOptions, LiveTranscriptionEvents

# Load environment variables
load_dotenv()

# Deepgram API Key
DEEPGRAM_API_KEY = os.getenv('DEEPGRAM_API_KEY')

# Claude WebSocket URL
CLAUDE_WS_URL = os.getenv('CLAUDE_WS_URL', 'ws://localhost:3000')

# Audio settings
CHANNELS = 1
FRAME_RATE = 16000
CHUNK = 8000
FORMAT = pyaudio.paInt16

# Global variables
transcript_buffer = ""
is_final = False

async def send_to_claude(message):
    async with websockets.connect(CLAUDE_WS_URL) as ws:
        await ws.send(json.dumps({
            "type": "send_message",
            "content": message
        }))
        response = await ws.recv()
        print(f"Claude response: {response}")

def on_message(self, result, **kwargs):
    global transcript_buffer, is_final
    sentence = result.channel.alternatives[0].transcript
    if len(sentence) == 0:
        return
    print(f"Transcription: {sentence}")
    transcript_buffer += sentence + " "
    if "done" in sentence.lower():
        is_final = True

def on_error(self, error, **kwargs):
    print(f"Error: {error}")

def audio_generator():
    p = pyaudio.PyAudio()
    stream = p.open(format=FORMAT, channels=CHANNELS, rate=FRAME_RATE, input=True, frames_per_buffer=CHUNK)
    
    try:
        while True:
            yield stream.read(CHUNK)
    finally:
        stream.stop_stream()
        stream.close()
        p.terminate()

def main():
    global is_final

    # Configure DeepgramClientOptions
    config = DeepgramClientOptions(
        options={"keepalive": "true"}
    )

    # Create a Deepgram client
    deepgram = DeepgramClient(DEEPGRAM_API_KEY, config)

    # Create a websocket connection
    dg_connection = deepgram.listen.live.v("1")

    # Register event handlers
    dg_connection.on(LiveTranscriptionEvents.Transcript, on_message)
    dg_connection.on(LiveTranscriptionEvents.Error, on_error)

    # Set up LiveOptions
    options = LiveOptions(
        punctuate=True,
        interim_results=False,
        language='en-US',
        model='nova-2'
    )

    # Start the connection
    dg_connection.start(options)

    print("Listening... Say 'done' to send the transcribed message to Claude.")

    try:
        for audio_chunk in audio_generator():
            if is_final:
                break
            dg_connection.send(audio_chunk)
    except KeyboardInterrupt:
        print("\nStopping...")
    finally:
        dg_connection.finish()

    if is_final:
        print(f"Sending message to Claude: {transcript_buffer.strip()}")
        asyncio.run(send_to_claude(transcript_buffer.strip()))

if __name__ == "__main__":
    main()

File: src/listen.ts
import { createClient, LiveTranscriptionEvents } from "@deepgram/sdk";
import { config } from "dotenv";
import fs from "fs";
import record from "node-record-lpcm16";
import WebSocket from "ws";
import { z } from "zod";

// Load environment variables
config();

// Define and validate environment variables
const envSchema = z.object({
  DEEPGRAM_API_KEY: z.string().min(1, {
    message:
      "DEEPGRAM_API_KEY is required. Get one at https://console.deepgram.com/signup?jump=keys",
  }),
  CLAUDE_WS_URL: z.string().url({
    message: "CLAUDE_WS_URL must be a valid WebSocket URL",
  }),
});

function log(message: string) {
  const timestamp = new Date().toISOString();
  const logMessage = `${timestamp}: ${message}\n`;
  console.log(logMessage);
  fs.appendFileSync("listen_log.txt", logMessage);
}

function logError(message: string, error: any) {
  const timestamp = new Date().toISOString();
  const errorMessage = `${timestamp} ERROR: ${message}\n${
    error.stack || error
  }\n`;
  console.error(errorMessage);
  fs.appendFileSync("listen_error_log.txt", errorMessage);
}

try {
  const env = envSchema.parse(process.env);

  // Initialize Deepgram client
  const deepgram = createClient(env.DEEPGRAM_API_KEY);

  // Initialize WebSocket connection to Claude
  const claudeWs = new WebSocket(env.CLAUDE_WS_URL);

  claudeWs.on("open", () => {
    log("Connected to Claude WebSocket");
  });

  claudeWs.on("error", (error) => {
    logError("Claude WebSocket error:", error);
  });

  // Start Deepgram live transcription
  const live = deepgram.listen.live({
    model: "nova",
    punctuate: true,
    language: "en-US",
    encoding: "linear16",
    sample_rate: 16000,
  });

  live.on(LiveTranscriptionEvents.Open, () => {
    log("Deepgram connection opened");

    // Start recording and streaming audio to Deepgram
    let recording;
    try {
      recording = record.record({
        sampleRate: 16000,
        channels: 1,
        audioType: "raw",
        recorder: "sox",
      });

      recording
        .stream()
        .on("data", (chunk) => {
          try {
            live.send(chunk);
          } catch (error) {
            logError("Error sending audio chunk to Deepgram:", error);
          }
        })
        .on("error", (error) => {
          logError("Error in audio stream:", error);
        });

      log("Recording started");
    } catch (error) {
      logError("Error starting audio recording:", error);
    }
  });

  let transcriptBuffer = "";

  live.on(LiveTranscriptionEvents.Transcript, (data) => {
    try {
      const transcript = data.channel.alternatives[0].transcript;
      if (transcript.trim()) {
        log(`Transcript: ${transcript}`);
        transcriptBuffer += transcript + " ";

        if (transcript.toLowerCase().includes("done")) {
          log(`Sending message to Claude: ${transcriptBuffer}`);
          claudeWs.send(
            JSON.stringify({
              type: "send_message",
              content: transcriptBuffer.trim(),
            })
          );
          transcriptBuffer = "";
        }
      }
    } catch (error) {
      logError("Error processing transcript:", error);
    }
  });

  live.on(LiveTranscriptionEvents.Error, (error) => {
    logError("Deepgram error:", error);
  });

  claudeWs.on("message", (data) => {
    try {
      const message = JSON.parse(data.toString());
      if (message.type === "message") {
        log(`Claude response: ${message.content}`);
      }
    } catch (error) {
      logError("Error processing Claude response:", error);
    }
  });

  process.on("SIGINT", () => {
    log("Shutting down...");
    live.finish();
    claudeWs.close();
    process.exit(0);
  });

  log('Listening... Say "done" to send the transcribed message to Claude.');
} catch (error) {
  logError("Error during setup:", error);
  process.exit(1);
}
